include/RAJA/pattern/kernel/ArgHelper.hpp

```diff
@@ -0,0 +1,293 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   Header file for kernel lambda executor.
+ *
+ ******************************************************************************
+ */
+
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC.
+//
+// Produced at the Lawrence Livermore National Laboratory
+//
+// LLNL-CODE-689114
+//
+// All rights reserved.
+//
+// This file is part of RAJA.
+//
+// For details about use and distribution, please read RAJA/LICENSE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+
+#ifndef RAJA_pattern_kernel_arghelper_HPP
+#define RAJA_pattern_kernel_arghelper_HPP
+
+#include "RAJA/config.hpp"
+
+#include <iostream>
+#include <type_traits>
+
+#include "camp/camp.hpp"
+#include "camp/concepts.hpp"
+#include "camp/tuple.hpp"
+
+#include "RAJA/util/macros.hpp"
+#include "RAJA/util/types.hpp"
+
+#include "RAJA/pattern/kernel/OffSet.hpp"
+#include "RAJA/pattern/kernel/Param.hpp"
+#include "RAJA/pattern/kernel/Seg.hpp"
+
+namespace RAJA
+{
+
+namespace internal
+{
+
+using RAJA::statement::SegList;
+using RAJA::statement::Seg;
+
+using RAJA::statement::OffSet;
+using RAJA::statement::OffSetList;
+
+using RAJA::statement::ParamList;
+using RAJA::statement::Param;
+
+//Extracts arguments from segments, and parameters
+template<typename Head, typename...Tail>
+struct extractor
+{};
+
+template<camp::idx_t id, typename...Tail>
+struct extractor<RAJA::statement::OffSet<id>, Tail...>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto extract_arg(Data &&data) ->
+    decltype(camp::get<id>(data.offset_tuple))
+  {
+    return camp::get<id>(data.offset_tuple);
+  }
+};
+
+template<camp::idx_t id, typename...Tail>
+struct extractor<RAJA::statement::Seg<id>, Tail...>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto extract_arg(Data &&data) ->
+    decltype(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)])
+  {
+    return camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)];
+  }
+};
+
+template<camp::idx_t id, typename...Tail>
+struct extractor<RAJA::statement::Param<id>, Tail...>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto extract_arg(Data &&data)->
+    decltype(camp::get<id>(data.param_tuple))
+  {
+    return camp::get<id>(data.param_tuple);
+  }
+};
+
+//Helper to concatante list [Move to camp?]
```

Comment:
camp::list is a type, and has no data.... so catList should be implemented using types not functions that return objects.

That is, the usage should look like:`catList<...>::type`
instead of  `catList<...>::makeList`

This will make implementation a lot cleaner.

Also, it may make sense to rename catList to catListHelper (maybe in an internal namespace) and then define:
`template<typename A, typename B>
using catList = typename internal::catListHelper<A,B>::type`

Reply:
Thanks! That is much better!